controllers/vspherevm_controller.go

```diff
@@ -306,6 +308,31 @@ func (r vmReconciler) reconcileDelete(ctx *context.VMContext) (reconcile.Result,
 		r.Logger.V(6).Info("unable to delete node", "err", err)
 	}
 
+	// Remove finalizers from any ipam claims
+	for devIdx, device := range ctx.VSphereVM.Spec.Network.Devices {
+		for poolRefIdx := range device.AddressesFromPools {
+			// check if claim exists
+			ipAddrClaim := &ipamv1.IPAddressClaim{}
+			ipAddrClaimName := fmt.Sprintf("%s-%d-%d", ctx.VSphereVM.Name, devIdx, poolRefIdx)
+			ctx.Logger.Info(fmt.Sprintf("removing finalizer for IPAddressClaim %s", ipAddrClaimName))
+			ipAddrClaimKey := apitypes.NamespacedName{
+				Namespace: ctx.VSphereVM.Namespace,
+				Name:      ipAddrClaimName,
+			}
+			if err := ctx.Client.Get(ctx, ipAddrClaimKey, ipAddrClaim); err != nil {
+				if apierrors.IsNotFound(err) {
+					continue
+				}
+				return reconcile.Result{}, errors.Wrapf(err, fmt.Sprintf("failed to find IPAddressClaim %s to remove the finalizer", ipAddrClaimName))
```

Comment:
This line should be called out. We see that other errors in the delete reconcile do not return errors, they log and eat the errors. We've chosen to return the error during finalizer removal errors.

Does anyone have context on why errors are not returned? Should this new code not return errors?

Reply:
@srm09 Do you have context on why during delete that errors are ignored and not returned?

Reply:
It depends on whether or not the resource should be immediately requeued for retry. If so, return the error. If you only want to retry during a normal reconciliation interval, eat the error and return nil.

Reply:
In our case, we need to make sure the claim's finalizer is removed, or the IP will not be returned to the pool, so we think it's appropriate to return an error.

It is odd to us that when deleting the VM object that if something were to go wrong, it would eat the error. That is why we ask if there's some other reason we should eat the error too.

Reply:
I don't think so. Probably historical mostly and trying to avoid early exponential backoff.